core/src/main/java/org/elasticsearch/search/internal/SearchContext.java

```diff
@@ -291,11 +291,18 @@ public InnerHitsContext innerHits() {
 
     public abstract SearchContext size(int size);
 
-    public abstract boolean hasFieldNames();
+    public abstract boolean hasStoredFields();
 
-    public abstract List<String> fieldNames();
+    public abstract boolean hasStoredFieldsContext();
 
-    public abstract void emptyFieldNames();
+    /**
+     * A shortcut function to see whether there is a storedFieldsContext and it says the fields are requested.
+     */
+    public abstract boolean storedFieldsRequested();
+
+    public abstract StoredFieldsContext storedFieldsContext();
+
+    public abstract SearchContext storedFieldsContext(StoredFieldsContext storedFieldsContext);
```

Comment:
They are replacements for hasFieldNames(), fieldNames() and emptyFieldNames().
I removed the stored fields context setter since it does not need to be in the abstract class.